compiler/src/iree/compiler/Dialect/HAL/IR/test/executable_ops.mlir

```diff
@@ -67,6 +68,47 @@ hal.executable @ex_with_workgroup_count_region {
 
 #executable_target_format = #hal.executable.target<"backend", "format">
 
+// CHECK-LABEL: @ex_with_condition
+hal.executable @ex_with_condition {
+  // CHECK: hal.executable.variant public @backend target(#executable_target_format
+  hal.executable.variant @backend target(#executable_target_format) {
+    // CHECK: hal.executable.condition(%[[DEVICE:.+]]: !hal.device) -> i1 {
+    hal.executable.condition(%device: !hal.device) -> i1 {
+      // CHECK-NEXT: %[[OK:.+]], %[[VALUE:.+]] = hal.device.query<%[[DEVICE]]
+      %ok, %value = hal.device.query<%device : !hal.device> key("some" :: "value") : i1, i32
+      // CHECK-NEXT: return %[[OK]]
+      hal.return %ok : i1
+    }
```

Comment:
So this could be used to check for groups of conditions, like having 3 different Vulkan extensions all available?

(query multiple times then AND those values somehow?)

Reply:
Yep! And device limits and such ("shared memory over 64k" etc). We memoize/hoist the queries so even if every dispatch does a query they all get done once at initialization time and stored in globals. There's more hoisting we can do though such that every dispatch ends up with a global indicating which variant it should use and then a branch table does the jump at runtime to the right routine (unless they all collapse to the same thing, which happens if ordinals line up/etc as they should from codegen results).